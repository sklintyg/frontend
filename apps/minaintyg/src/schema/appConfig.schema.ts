import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
