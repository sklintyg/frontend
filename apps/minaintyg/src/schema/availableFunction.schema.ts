import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
