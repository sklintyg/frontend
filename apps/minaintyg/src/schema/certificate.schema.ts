import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
