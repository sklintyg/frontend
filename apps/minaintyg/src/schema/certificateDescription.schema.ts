import { z } from 'zod'

export const certificateDescriptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  detailedDescription: z.string(),
  issuerTypeId: z.string(),
  cssPath: z.string(),
  scriptPath: z.string(),
  dependencyDefinitionPath: z.string(),
  defaultRecipient: z.string(),
  deprecated: z.boolean(),
  displayDeprecated: z.boolean(),
})

export type CertificateDescription = z.infer<typeof certificateDescriptionSchema>
