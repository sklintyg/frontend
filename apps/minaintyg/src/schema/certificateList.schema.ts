import { z } from 'zod'

export enum CertificateStatus {
  NEW = 'Nytt',
  REPLACED = 'Ers√§tter intyg',
  SENT = 'Skickat',
  NOT_SENT = 'Ej skickat',
}

export const certificateListSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateListEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateListIssuerSchema = z.object({ name: z.string() })
export const certificateUnitSchema = z.object({ id: z.string(), name: z.string() })
export const certificateTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  version: z.string(),
})

export const certificateListItemSchema = z.object({
  summary: certificateListSummarySchema,
  issuer: certificateListIssuerSchema,
  unit: certificateUnitSchema,
  events: z.array(certificateListEventSchema),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
})

export type CertificateListEvent = z.infer<typeof certificateListEventSchema>
export type CertificateListIssuer = z.infer<typeof certificateListIssuerSchema>
export type CertificateListUnit = z.infer<typeof certificateUnitSchema>
export type CertificateListSummary = z.infer<typeof certificateListSummarySchema>
export type CertificateListItem = z.infer<typeof certificateListItemSchema>
