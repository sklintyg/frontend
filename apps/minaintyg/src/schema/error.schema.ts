import { z } from 'zod'

export const ErrorType = z.enum(['unknown', 'login-failed', 'unavailable', 'logged-out', 'not-found', 'underage'])
export type ErrorTypeEnum = z.infer<typeof ErrorType>

export const ErrorPageActionType = z.enum(['start', 'login', '1177', 'about'])
export type ErrorPageActionTypeEnum = z.infer<typeof ErrorPageActionType>

export const ErrorCode = z.enum([
  'CLIENT_ERROR',
  'LOGIN_FAILED',
  'UNKNOWN_INTERNAL_ERROR',
  'FETCH_ERROR',
  'PARSING_ERROR',
  'TIMEOUT_ERROR',
  'CUSTOM_ERROR',
])

export type ErrorCodeEnum = z.infer<typeof ErrorCode>

export const errorSchema = z.object({
  id: z.string(),
  code: ErrorCode.or(z.number()),
  message: z.string(),
  stackTrace: z.string().nullable().optional(),
})

export type ErrorData = z.infer<typeof errorSchema>
