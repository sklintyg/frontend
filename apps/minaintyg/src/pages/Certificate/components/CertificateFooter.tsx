import { CertificateMetadata } from '../../../schema/certificate.schema'

export function CertificateFooter({ issuer, unit }: CertificateMetadata) {
  return (
    <footer className="border-stone-clear my-5 border-y py-5">
      <h2 className="ids-heading-3">Intyget är utfärdat och signerat av</h2>
      <address className="flex flex-col gap-4 not-italic md:flex-row md:gap-[6.25rem]">
        <div>
          <strong>{issuer.name}</strong>
          <p>{issuer.phoneNumber}</p>
        </div>
        <div>
          <strong>{unit.name}</strong>
          <p>{unit.address}</p>
        </div>
      </address>
    </footer>
  )
}
