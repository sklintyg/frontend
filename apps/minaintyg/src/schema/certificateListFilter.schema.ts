import { z } from 'zod'
import { CertificateStatus } from './certificateList.schema'

export const certificateFilterOptionsSchema = z.object({
  total: z.number(),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  units: z.array(z.string()),
  certificateTypes: z.array(z.string()),
  years: z.array(z.string()),
})

type SelectedOptions<T> = { [X in keyof T]: T[X] extends Array<unknown> ? T[X][number] : T[X] }

export type CertificateFilterOptions = z.infer<typeof certificateFilterOptionsSchema>
export type CertificateSelectedOptions = SelectedOptions<CertificateFilterOptions>
