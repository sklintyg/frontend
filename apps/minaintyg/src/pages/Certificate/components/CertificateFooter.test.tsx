import { fakerFromSchema } from '@frontend/fake'
import { render, screen } from '@testing-library/react'
import { certificateMetadataSchema } from '../../../schema/certificate.schema'
import { CertificateFooter } from './CertificateFooter'

const metadata = fakerFromSchema(certificateMetadataSchema)({
  issuer: {
    name: 'Arnold Johansson',
  },
  unit: {
    name: 'Alfa-enheten, Alfamottagningen',
    address: 'Gatuadressens väg 1',
    phoneNumber: '08-123 456 78',
    city: 'Storstaden',
    zipCode: '111 11',
  },
})
it('Should render issuerName', () => {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    issuer: {
      name: 'Arnold Johansson',
    },
  })

  render(<CertificateFooter {...metadata} />)
  expect(screen.getByText(`Arnold Johansson`)).toBeInTheDocument()
})

it('Should render unit phoneNumber', () => {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    unit: {
      phoneNumber: '08-123 456 78',
    },
  })

  render(<CertificateFooter {...metadata} />)
  expect(screen.getByText(`08-123 456 78`)).toBeInTheDocument()
})

it('Should only render unit name if matching careUnit name', () => {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    unit: {
      name: 'Alfa-enheten',
    },
    careUnit: {
      name: 'Alfa-enheten',
    },
  })

  render(<CertificateFooter {...metadata} />)
  expect(screen.getByText(`Alfa-enheten`)).toBeInTheDocument()
})

it('Should render unit name and careUnit name if not the same', () => {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    unit: {
      name: 'Alfa-enheten',
    },
    careUnit: {
      name: 'Alfamottagningen',
    },
  })

  render(<CertificateFooter {...metadata} />)
  expect(screen.getByText(`Alfa-enheten, Alfamottagningen`)).toBeInTheDocument()
})

it('Should render unit address', () => {
  const metadata = fakerFromSchema(certificateMetadataSchema)({
    unit: {
      address: 'Gatuadressens väg 1',
      city: 'Storstaden',
      zipCode: '111 11',
    },
  })

  render(<CertificateFooter {...metadata} />)
  expect(screen.getByText(`Gatuadressens väg 1, 111 11, Storstaden`)).toBeInTheDocument()
})
