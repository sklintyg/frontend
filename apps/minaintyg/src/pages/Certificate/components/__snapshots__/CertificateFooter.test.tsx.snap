// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<div>
  <footer>
    <hr
      class="border-stone-clear my-7"
    />
    <h2
      class="ids-heading-3"
    >
      Intyget är utfärdat och signerat av
    </h2>
    <address
      class="flex flex-col gap-4 not-italic md:flex-row md:gap-[6.25rem]"
    >
      <div>
        <strong>
          Arnold Johansson
        </strong>
        <p>
          08-123 456 78
        </p>
      </div>
      <div>
        <strong>
          Alfa-enheten, Alfamottagningen
        </strong>
        <p>
          Gatuadressens väg 1, 111 11 Storstaden
        </p>
      </div>
    </address>
    <hr
      class="border-stone-clear my-7"
    />
  </footer>
</div>
`;
